### PR TITLE
Associations

### PR DESCRIPTION
**Pull request by rachitchaurasia99**
Associations

We have before_destroy :ensure_not_referenced_by_line_item in Product. Now lets try a better implementation of this  using association options. So a product should not be destroyed if there is any line_item(s) associated with the product.
 Create an association on cart through which we could get all the products associated with the cart.
 Create an association on product through which we could get all the carts associated with the product.
 Create an association on cart model through which I could get all enabled products associated with it.

 Everytime a line_item is added or removed from cart, line_items_count column in cart table should be automatically incremented or decremented	.

Add an association between user and order. Then add another page in the application which can be accessed by localhost:3000/users/orders and display logged in user’s all orders with total amount and line_items. 

Then add another page localhost:3000/users/line_Items which should display all line_items that current users has ordered till date. All lineitems for the user should come from association. like user.line_items
This list should be paginated. 5 items per page

### CODE REVIEW

File: app/models/product.rb [hunk 1]
```diff
@@ -1,6 +1,7 @@
 class Product < ApplicationRecord
-  has_many :line_items
+  has_many :line_items, dependent: :restrict_with_exception
```
Comment (siddharth-vinsol) [inline app/models/product.rb]: which is a better option here
`restrict_with_exception` or `restrict_with_error`

File: app/models/line_item.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 class LineItem < ApplicationRecord
   belongs_to :order, optional: true
   belongs_to :product
-  belongs_to :cart, optional: true
+  belongs_to :cart, optional: true, counter_cache: true
```
Comment (siddharth-vinsol) [inline app/models/line_item.rb]: What will happend if `carts` table does now have `line_items_count` column

File: app/views/users/line_item.erb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  <h3>Line Item#<%= counter = counter + 1 %></h3>
+  <br>
+  product: <%= line_item.product.title %>
+  <br>
+  quantity: <%= line_item.quantity %>
+  <br>
+  order_id: <%= line_item.order_id %>
+  <br>
+  <br>
+<% end %>
```
Comment (siddharth-vinsol) [inline app/views/users/line_item.erb]: add empty line after this

File: app/views/users/orders.html.erb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  <h2>ORDER#<%= counter = counter+1 %></h2>
+  customer: <%= order.name %>
+  <br>
+  email: <%= order.email %>
+  <br>
+  <h3>line items:</h3>
+  <br>
+  <% total_order_price = 0 %>
+  <% order.line_items.each do |line_item| %>
+    <%= Product.find(line_item.product_id).title %>
```
Comment (siddharth-vinsol) [inline app/views/users/orders.html.erb]: why not use association on line_item for product here

File: app/views/users/line_item.erb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<% counter = 0 %>
+<% @user.line_items.each do |line_item| %>
```
Comment (siddharth-vinsol) [inline app/views/users/line_item.erb]: use better way here

File: app/views/users/line_item.erb
Comment (siddharth-vinsol) [inline app/views/users/line_item.erb:1]: Add pagination here as state in exercise

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -19,6 +19,14 @@ def new
...
   end
 
+  def orders
+    @user = User.find(session[:user_id])
+  end
+
+  def line_item
+    @user = User.find(session[:user_id])
+  end
+
```
Comment (siddharth-vinsol) [inline app/controllers/users_controller.rb]: 1. `@user = User.find(session[:user_id])` is repeated code here so extract this out at better place where multiple controllers can access it
2. Why not here only load order and line_item so it will not be responsibilty of view to load it

File: app/models/product.rb [hunk 7]
```diff
@@ -1,6 +1,7 @@
...
   has_many :orders, through: :line_items
+  has_many :carts, through: :line_items
```
Comment (siddharth-vinsol) [inline app/models/product.rb:4]: why this

**Review by siddharth-vinsol — CHANGES_REQUESTED** [review]

File: app/models/product.rb (hunk 7 above)
Comment (rachitchaurasia99) [reply app/models/product.rb:4]: To get all carts associated with product and since they are linked via line_items, so we use ```through:``` option here.

File: app/models/product.rb (hunk 1 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: ```restrict_with_error``` as in this case there is no need to handle exception.

File: app/views/users/line_item.erb (hunk 3 above)
Comment (rachitchaurasia99) [reply app/views/users/line_item.erb]: done

File: app/views/users/orders.html.erb (hunk 4 above)
Comment (rachitchaurasia99) [reply app/views/users/orders.html.erb]: corrected

File: app/controllers/users_controller.rb (hunk 6 above)
Comment (rachitchaurasia99) [reply app/controllers/users_controller.rb]: done

File: app/views/users/line_item.erb (hunk 5 above)
Comment (rachitchaurasia99) [reply app/views/users/line_item.erb]: done

File: app/models/line_item.rb (hunk 2 above)
Comment (rachitchaurasia99) [reply app/models/line_item.rb]: According to the convention counter_cache will try to find ```line_items_count``` column, so since the table doesn't have it, it will give Attribute Error.

File: app/views/users/line_item.erb
Comment (rachitchaurasia99) [reply app/views/users/line_item.erb:1]: done

File: db/schema.rb
Comment (rachitchaurasia99) [inline db/schema.rb:1]: Have to drop the database and recreate.

**Comment by apekshit-tewathia** [comment]
@Ratchet1999 please reply to all the comments in all the PRs
Especially when questions are asked.

**Comment by rachitchaurasia99** [comment]
> @Ratchet1999 please reply to all the comments in all the PRs Especially when questions are asked.

ok sir

File: app/controllers/concerns/current_user.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+module CurrentUser
+  private
+
+    def current_user
+      User.find(session[:user_id])
+    end
+end
```
Comment (siddharth-vinsol) [inline app/controllers/concerns/current_user.rb]: What if this method is called 5 times in one controller action
Are we going to make same query 5 times

File: app/controllers/orders_controller.rb [hunk 9]
```diff
@@ -1,5 +1,5 @@
 class OrdersController < ApplicationController
-  include CurrentCart
+  include CurrentCart, CurrentUser
```
Comment (siddharth-vinsol) [inline app/controllers/orders_controller.rb]: Just for readability and avoid confusion\

```suggestion
  include CurrentUser
  include CurrentCart
```

File: app/controllers/line_items_controller.rb [hunk 10]
```diff
@@ -9,7 +9,7 @@ class LineItemsController < ApplicationController
   def create
     product = Product.find(params[:product_id])
     @line_item = @cart.add_product(product)
-    @line_item = @cart.line_items.build(product: product)
+
```
Comment (siddharth-vinsol) [inline app/controllers/line_items_controller.rb:12]: why this change?

File: app/helpers/line_items_helper.rb [hunk 11]
```diff
@@ -0,0 +1,5 @@
+module LineItemsHelper
+  def total_price(line_items)
+    line_items.to_a.sum { |item| item.total_price }
```
Comment (siddharth-vinsol) [inline app/helpers/line_items_helper.rb]: Can we use SQL query instead of converting this to ruby array and processing that

File: app/controllers/orders_controller.rb [hunk 12]
```diff
@@ -27,8 +27,7 @@ def edit
   def create
     @order = Order.new(order_params)
     @order.add_line_items_from_cart(@cart)
-
-    @order.save!
+    @order.user_id = current_user.id
```
Comment (siddharth-vinsol) [inline app/controllers/orders_controller.rb]: can we assign user_id when builidng the order in line 28

File: app/models/line_item.rb [hunk 13]
```diff
@@ -3,7 +3,9 @@ class LineItem < ApplicationRecord
   belongs_to :product
   belongs_to :cart, optional: true
   
-  validates :product_id, uniqueness: { scope: :cart_id }
+  validates :product_id, uniqueness: { scope: :cart_id }, if: :cart
+  
+  paginates_per 5
```
Comment (siddharth-vinsol) [inline app/models/line_item.rb]: `5` here should be extracted in constant in this class only

File: app/models/order.rb [hunk 14]
```diff
@@ -6,6 +6,7 @@ class Order < ApplicationRecord
     "Purchase order" => 2
   }
   
+  belongs_to :user
```
Comment (siddharth-vinsol) [inline app/models/order.rb:9]: should there be any extra configuration here like `dependent` or something else

File: app/views/line_items/_line_item.html.erb [hunk 15]
```diff
@@ -6,4 +6,4 @@
 <td class="quantity"><%= line_item.quantity %></td>
 <td><%= line_item.product.title %></td>
 <td class="price"><%= number_to_currency(line_item.total_price) %></td>
-</tr>
\ No newline at end of file
+</tr>
```
Comment (siddharth-vinsol) [inline app/views/line_items/_line_item.html.erb]: fix identation this file

File: config/routes.rb [hunk 16]
```diff
@@ -9,6 +9,9 @@
   get 'sessions/create'
   get 'sessions/destroy'
 
+  get 'users/line_items', to: 'users#line_items'
+  get 'users/orders', to: 'users#orders'
```
Comment (siddharth-vinsol) [inline config/routes.rb]: Can we write code that assign `controller` to both without repeating `users#`

**Review by siddharth-vinsol — CHANGES_REQUESTED** [review]

File: app/controllers/line_items_controller.rb (hunk 10 above)
Comment (rachitchaurasia99) [reply app/controllers/line_items_controller.rb:12]: was trying something and forgot to undo

File: app/models/order.rb (hunk 14 above)
Comment (rachitchaurasia99) [reply app/models/order.rb:9]: No, as there is already a dependent option in user.rb for ```has_many :orders``` . Having dependent on both sides  an lead to orphaned records in the database.

File: app/controllers/concerns/current_user.rb (hunk 8 above)
Comment (rachitchaurasia99) [reply app/controllers/concerns/current_user.rb]: done

File: app/controllers/orders_controller.rb (hunk 9 above)
Comment (rachitchaurasia99) [reply app/controllers/orders_controller.rb]: done

File: app/helpers/line_items_helper.rb (hunk 11 above)
Comment (rachitchaurasia99) [reply app/helpers/line_items_helper.rb]: done

File: app/controllers/orders_controller.rb (hunk 12 above)
Comment (rachitchaurasia99) [reply app/controllers/orders_controller.rb]: Yes. we can if we create new order using ```build ``` method with current user object

File: app/models/line_item.rb (hunk 13 above)
Comment (rachitchaurasia99) [reply app/models/line_item.rb]: done

File: app/views/line_items/_line_item.html.erb (hunk 15 above)
Comment (rachitchaurasia99) [reply app/views/line_items/_line_item.html.erb]: done

File: config/routes.rb (hunk 16 above)
Comment (rachitchaurasia99) [reply config/routes.rb]: done

File: app/helpers/line_items_helper.rb
Comment (siddharth-vinsol) [inline app/helpers/line_items_helper.rb:1]: Why this if not used anywhere

File: app/models/line_item.rb [hunk 17]
```diff
@@ -5,6 +5,8 @@ class LineItem < ApplicationRecord
   
   validates :product_id, uniqueness: { scope: :cart_id }, if: :cart
   
+  paginates_per DEFAULT_NUMBER_OF_PAGES
```
Comment (siddharth-vinsol) [inline app/models/line_item.rb]: Rename constant name

File: config/initializers/constants.rb [hunk 18]
```diff
@@ -3,3 +3,5 @@
 DESCRIPTION_REGEX = /\A(\w+ ){4,9}\w+ *\z/
 MINIMUM_PRICE = 0.01
 ADMIN_EMAIL = 'admin@depot.com'
+DEFAULT_NUMBER_OF_PAGES = 5
```
Comment (siddharth-vinsol) [inline config/initializers/constants.rb]: This shouldnt be defined here

File: test/mailers/previews/user_mailer_preview.rb
Comment (siddharth-vinsol) [inline test/mailers/previews/user_mailer_preview.rb:1]: Why this file added when nothing related to user mailer done in this PR

**Review by siddharth-vinsol — CHANGES_REQUESTED** [review]

File: app/helpers/line_items_helper.rb
Comment (rachitchaurasia99) [reply app/helpers/line_items_helper.rb:1]: fixed

File: app/models/line_item.rb (hunk 17 above)
Comment (rachitchaurasia99) [reply app/models/line_item.rb]: done

File: config/initializers/constants.rb (hunk 18 above)
Comment (rachitchaurasia99) [reply config/initializers/constants.rb]: done

File: test/mailers/previews/user_mailer_preview.rb
Comment (rachitchaurasia99) [reply test/mailers/previews/user_mailer_preview.rb:1]: fixed

**Review by siddharth-vinsol — APPROVED** [review]
